Implement SqlException.IsTransient

tbolon opened the pull request:
Fix #649

Code is based (it's a simple copy/paste for now) on [EFCore SqlServerTransientExceptionDetector.cs](https://github.com/dotnet/efcore/blob/main/src/EFCore.SqlServer/Storage/Internal/SqlServerTransientExceptionDetector.cs)

The xmldoc is quite empty (and does not respect the convention), I expect some help and/or guidance about how it should be handled: should we document all the cases handled by the code? Or keep it as a black box? Or something between where the a direct link to the code is added in the documentation?

Should we move the code outside the class to keep the file size small?

Should we expose this algorithm outside (make the external class public) so other libraries can use it?

tbolon commented:
Ping @AndriySvyryd which has handled most of the work on dotnet/efcore.

tbolon commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlException.cs:
```diff
@@ -99,6 +99,204 @@ public override void GetObjectData(SerializationInfo si, StreamingContext contex
+                    }
+                }
+
+                return false;
```
You think it is not? In which cases there are multiple errors returned for an exception? Anyway, if we retry because there is at least one transient error, and that error disappear, we can perhaps expect that other non-transient error will disappear. If we fail early we will not be able to detect if they are related.

roji replied:
> Anyway, if we retry because there is at least one transient error, and that error disappear, we can perhaps expect that other non-transient error will disappear. If we fail early we will not be able to detect if they are related.

I'm not very knowledgeable about SQL Server/SqlClient error handling (or which situation produces multiple errors), but generally speaking, the whole point of an error being non-transient is that it doesn't just go away when you retry the same operation. So I don't think it makes sense for IsTransient to return true when there's at least one non-transient error.

tbolon replied:
I have reverted the check.

cheenamalhotra replied:
The retry in the driver is based on whether any of the multiple errors is transient. So I believe it would be fair to return `IsTransient` based on the same condition check.

ErikEJ commented:
There already is a list of transient errors in the codebase, let's stick to this one: https://github.com/dotnet/SqlClient/blob/c7cefee758f02cf0127f448e41e5c2f02022a062/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Reliability/SqlConfigurableRetryFactory.cs#L40

tbolon commented:
> There already is a list of transient errors in the codebase, let's stick to this one:
> 
> https://github.com/dotnet/SqlClient/blob/c7cefee758f02cf0127f448e41e5c2f02022a062/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Reliability/SqlConfigurableRetryFactory.cs#L40

What could be the prefered way to reuse that list? Add an internal method that I could call from the property or move out the list of codes in a specific class?

log-trader commented on doc/snippets/Microsoft.Data.SqlClient/SqlException.xml:
```diff
@@ -153,6 +153,10 @@ catch (Exception ex) {
         </GetObjectData>
+        <IsTransient>
+            <summary>Indicates whether the error represented by this <see cref="T:Microsoft.Data.SqlClient.SqlException" /> could be a transient error, i.e. if retrying the triggering operation may succeed without any other change. Examples of transient errors include failure to acquire a database lock, networking issues. This allows automatic retry execution strategies to be developed without kn ...
+            <returns><see langword="true"/> if the error represented could be a transient error; <see langword="true"/> otherwise.</returns>
```
```suggestion
            <returns><see langword="true"/> if the error represented could be a transient error; <see langword="false"/> otherwise.</returns>
```

tbolon commented:
Ping @ErikEJ to resurect this PR.

stevendarby commented:
> My 2c would be that the list of errors that @ErikEJ referred to is considerably smaller than the list of errors handled by the EF code [...] On the face of it, the EF list is the better choice as it's more exhaustive

Just to add to this, the EF code has been updated very recently, while the SqlClient list is 3 years old. Seems much better to go with the former.

Importantly, the most recent addition to the EF switch statement does a check on more than just the number, so a HashSet of ints isn't going to cut it really.

@tbolon probably worth updating this PR to include the latest changes referred to above :)
https://github.com/dotnet/efcore/blob/main/src/EFCore.SqlServer/Storage/Internal/SqlServerTransientExceptionDetector.cs

cheenamalhotra commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlException.cs:
```diff
@@ -99,6 +99,207 @@ public override void GetObjectData(SerializationInfo si, StreamingContext contex
+                if (Errors.Count == 0)
+                    return false;
+
+                foreach (SqlError err in Errors)
```
Instead of this, which is a new list, would recommend using existing logic to determine whether the exception is going to be retried on.

cheenamalhotra commented:
While https://github.com/dotnet/SqlClient/issues/649 is open, this PR doesn't meet the design criteria and would need to be redone. Closing as stale and change not acceptable at the moment. A new PR for https://github.com/dotnet/SqlClient/issues/649 would be acceptable if is designed based on existing driver logic to determine transient errors.

stevendarby commented:
@cheenamalhotra the issue you would accept a PR for explicitly mentions that the EF approach could be used as starting point, and above I point out how the EF list more frequently maintained. What exactly are the design criteria and why?

cheenamalhotra commented:
Even though the issue mentions EF Core approach could be used, it doesn't mean after integration in the driver that continues to be the final approach. As a driver maintainer, I'm against duplicating error numbers for the sake of this API, and maintaining list of transient errors twice, and therefore not accepting this change.

stevendarby commented:
@cheenamalhotra 

From @roji's issue:

"The goal would eventually be for EF Core to rely on the new IsTransient property rather than contain specific knowledge on SqlClient error codes."

EF Core will not be able to use IsTransient if its implementation is poorer than what EF Core already has. Some coordination between Microsoft teams would be great to see here.

tbolon commented:
I also think that SqlException should be the reference for any other implementors. There will certainly be a period where both algorithms will require some maintenance (or *just* coordination). But in the end, once the IsTransient property will be known as available, all others implementations will change to be based on it.

Maybe the method I implemented as Internal in #2894 should be made public (and perhaps also moved as a static method in SqlError for example?) for other implementations to be able to reuse it (and detect it's available when targeting a recent version of MSSqlClient).
